identity-and-access-management/key-manager/how-to/use-tink-go.mdx

```diff
@@ -0,0 +1,310 @@
+
+Here is a way we could encrypt the data with AD before inserting it:
+
+```go
+handle, _ := keyset.ReadWithAssociatedData(r, kekAEAD, dekAD)
+primitive, _ := aead.New(handle)  // Same DEK for the two payloads
+
+secret1, _ := primitive.Encrypt([]byte("This is a secret"), []byte("id42"))
+// Insert secret1 into row 42
+
+secret2, _ := primitive.Encrypt([]byte("This is another secret"), []byte("id43"))
+// Insert secret2 into row 43
+```
+
+You should preferably use AD in general. AD does not need to be stored, as
+it can be infered from the context at decryption time.
+
+It is also possible to use a unique DEK for each payload.
+
+
+## Hierarchy of Keys
+
+Unlike KEKs that reside and are managed by Scaleway's KM, DEKs are "free":
+you can generate and have as much as you want.
+However, your application still needs to call the KM:
+ - At least once to encrypt a newly generated DEK before storing it, and
+ - Each time a DEK needs to be decrypted before use
+
+That would incur charges and make your application slow due to numerous calls to
+the Key Manager's API.
+
+Fortunately, Tink makes easy to work with arbitraty key hierarchy. In the following
+example, a DEK Master Key protects all the DEKs, while being protected by the remote
+KEK.
+
+```go
+// DEK Master Key (decrypted by the remote KEK)
+masterKeyFile, _ := os.Open("/path/to/encrypted_masterkey.tink")
+masterKeyHandle, _ := keyset.ReadWithAssociatedData(masterKeyFile, kekAEAD, []byte("master-key-001")) // Call the API
+masterKeyAEAD, _ := aead.New(masterKeyHandle)
+
+// DEK #1 (decrypted by the Master Key)
+dek1File, _ := os.Open("/path/to/encrypted_dek1.tink")
+dekHandle1, _ := keyset.ReadWithAssociatedData(dek1File, masterKeyAEAD, []byte("dek1"))  // No call to the API
+
+// DEK #2 (decrypted by the Master Key)
+dek2File, _ := os.Open("/path/to/encrypted_dek2.tink")
+dekHandle2, _ := keyset.ReadWithAssociatedData(dek2File, masterKeyAEAD, []byte("dek2")) // No call to the API
+
+
+// Encrypt and decrypt data
+
+dek1AEAD, _ := aead.New(dekHandle1)
+dek2AEAD, _ := aead.New(dekHandle2)
+
+ct1, _ := dek1AEAD.Encrypt([]byte("this is a secret"), []byte("id42"))
+ct2, _ := dek2AEAD.Encrypt([]byte("this is another secret"), []byte("id43"))
+
+/* ... */
+/* ... */
+```
+
+<Message type="note">
+You don't need to use the same algorithm and cipher between a key and its
+protecting key.
+</Message>
+
+The last hierarchy can work for most use-cases. However,
+there exists no "one fits all" approach to make the right key hierarchy. It is
+up to you to make a hierarchy that suits you best, according to your application
+needs and constraints.
+
+
+<Message type="note">
+Keys, either managed by the Key Manager or not, are not DEKs nor KEKs <i>per se</i>.
+It is the context in which you use them that makes them DEKs or KEKs. We usually
+assume that keys living in the Key Manager are only used to protect other keys, so they
+are usually called "KEKs" with this asusmption in mind.
+</Message>
```

Comment:
```suggestion
Unlike KEKs that reside and are managed by Key Manager, DEKs are free: you can generate and have as many as you want.

However, your application still needs to call the Key Manager API:

 - At least once to encrypt a newly generated DEK before storing it, and
 - Each time a DEK needs to be decrypted before use

Thus, you can use a hierarchy of keys to minimize calls to the Key Manager API (or any remote key mangement service), which can slow down your application and incur charges.

In the example below, the application only needs to call Key Manager once to decrypt the DEK Master Key. All subsequent decryption of DEKs happens locally, which improves efficiency.

```go
// The DEK Master Key (which protects all other DEKs) is stored and protected by the remote KEK.
masterKeyFile, _ := os.Open("/path/to/encrypted_masterkey.tink")
masterKeyHandle, _ := keyset.ReadWithAssociatedData(masterKeyFile, kekAEAD, []byte("master-key-001")) // Call the API
masterKeyAEAD, _ := aead.New(masterKeyHandle)

// DEK #1 encrypted and decrypted by the Master Key, rather than by the remote KEK directly. This avoids frequent API calls.
dek1File, _ := os.Open("/path/to/encrypted_dek1.tink")
dekHandle1, _ := keyset.ReadWithAssociatedData(dek1File, masterKeyAEAD, []byte("dek1"))  // No call to the API

// DEK #2 encrypted and decrypted by the Master Key, rather than by the remote KEK directly. This avoids frequent API calls.
dek2File, _ := os.Open("/path/to/encrypted_dek2.tink")
dekHandle2, _ := keyset.ReadWithAssociatedData(dek2File, masterKeyAEAD, []byte("dek2")) // No call to the API


// DEKs are used to encrypt and decrypt the actual data

dek1AEAD, _ := aead.New(dekHandle1)
dek2AEAD, _ := aead.New(dekHandle2)

ct1, _ := dek1AEAD.Encrypt([]byte("this is a secret"), []byte("id42"))
ct2, _ := dek2AEAD.Encrypt([]byte("this is another secret"), []byte("id43"))

/* ... */
/* ... */
```

<Message type="tip">
 Your DEK and KEK do not need to use the same algorithm and cipher.
</Message>

The example above can work for most use cases. However, there is no "one fits all" approach to create the right key hierarchy. It is up to you to decide on a hierarchy that suits you best,according to your application needs and constraints.


<Message type="important">
 Scaleway does not define keys managed by Key Manager (or any other key management service) as DEKs or KEKs. The context in which you use these keys makes them DEKs or KEKs. We usually assume that keys in Key Manager are only used to protect other keys, hence the use of the terms "KEKs" and "DEKs".
</Message>
```